File: app/src/main/java/wjy/yo/ereader/entity/UserData.java
package wjy.yo.ereader.entity;

import android.arch.persistence.room.ForeignKey;

import java.util.Objects;

import wjy.yo.ereader.entity.userdata.User;

public class UserData extends FetchedData {

    @ForeignKey(entity = User.class, parentColumns = "name", childColumns = "userName")
    protected String userName;

    protected boolean local;

    public String getUserName() {
        return userName;
    }

    public void setUserName(String userName) {
        this.userName = userName;
    }

    public boolean isLocal() {
        return local;
    }

    public void setLocal(boolean local) {
        this.local = local;
    }

    public boolean equals(Object o) {
        if (o == this) {
            return true;
        }
        if (!(o instanceof UserData)) {
            return false;
        }
        UserData other = (UserData) o;
        return Objects.equals(userName, other.userName) && local == other.local;
    }

}

File: app/src/main/java/wjy/yo/ereader/entity/UserData.kt
package wjy.yo.ereader.entity

import android.arch.persistence.room.ForeignKey

import wjy.yo.ereader.entity.userdata.User

open class UserData : FetchedData("") {

    @ForeignKey(entity = User::class, parentColumns = arrayOf("name"), childColumns = arrayOf("userName"))
    var userName: String? = null

    var isLocal: Boolean = false

    override fun equals(other: Any?): Boolean {
        if (other === this) {
            return true
        }
        if (other !is UserData) {
            return false
        }
        val otherUD = other as UserData?
        return userName == otherUD!!.userName && isLocal == otherUD.isLocal
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + (userName?.hashCode() ?: 0)
        result = 31 * result + isLocal.hashCode()
        return result
    }

}
